Mathlib/Algebra/Ring/Ext.lean

```diff
@@ -0,0 +1,528 @@
+semiring, ring, extensionality
+-/
+
+local macro:65 lhs:term:65 "+[" type:term ", " inst:term "]" rhs:term:66 : term =>
+  `(term| (letI := $inst; ($lhs : $type) + ($rhs : $type) : $type))
+local macro:70 lhs:term:70 "*[" type:term ", " inst:term "]" rhs:term:71 : term =>
+  `(term| (letI := $inst; ($lhs : $type) * ($rhs : $type) : $type))
+
+universe u
+
+variable {R : Type u}
+
+/-! ### Distrib -/
+namespace Distrib
+
+@[ext] theorem ext ⦃inst₁ inst₂ : Distrib R⦄
+    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+    (h_mul : ∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :
+    inst₁ = inst₂ := by
+  -- Split into `add` and `mul` functions and properties.
+  rcases inst₁ with @⟨⟨⟩, ⟨⟩⟩
+  rcases inst₂ with @⟨⟨⟩, ⟨⟩⟩
+  -- Prove equality of parts using function extensionality.
+  congr <;> (apply funext₂; assumption)
+
+theorem ext_iff (inst₁ inst₂ : Distrib R) :
+    inst₁ = inst₂ ↔
+      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+      ∧ (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
+  ⟨fun h ↦ by constructor <;> (intros; congr), And.elim (ext · ·)⟩
+
+end Distrib
+
+/-! ### NonUnitalNonAssocSemiring -/
+namespace NonUnitalNonAssocSemiring
+
+@[ext] theorem ext ⦃inst₁ inst₂ : NonUnitalNonAssocSemiring R⦄
+    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+    (h_mul : ∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :
+    inst₁ = inst₂ := by
+  -- Split into `AddMonoid` instance, `mul` function and properties.
+  rcases inst₁ with @⟨_, ⟨⟩⟩
+  rcases inst₂ with @⟨_, ⟨⟩⟩
+  -- Prove equality of parts using already-proved extensionality lemmas.
+  congr <;> (ext; apply_assumption)
+
+theorem toDistrib_injective : Function.Injective (@toDistrib R) := by
+  intro _ _ h
+  ext x y
+  · exact congrArg (·.toAdd.add x y) h
+  · exact congrArg (·.toMul.mul x y) h
+
+theorem ext_iff (inst₁ inst₂ : NonUnitalNonAssocSemiring R) :
+    inst₁ = inst₂ ↔
+      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+      ∧ (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
+  ⟨fun h ↦ by constructor <;> (intros; congr), And.elim (ext · ·)⟩
+
+end NonUnitalNonAssocSemiring
+
+/-! ### NonUnitalSemiring -/
+namespace NonUnitalSemiring
+
+theorem toNonUnitalNonAssocSemiring_injective :
+    Function.Injective (@toNonUnitalNonAssocSemiring R) := by
+  rintro ⟨⟩ ⟨⟩ _; congr
+
+@[ext] theorem ext ⦃inst₁ inst₂ : NonUnitalSemiring R⦄
+    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+    (h_mul : ∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :
+    inst₁ = inst₂ :=
+  toNonUnitalNonAssocSemiring_injective <|
+    NonUnitalNonAssocSemiring.ext h_add h_mul
+
+theorem ext_iff (inst₁ inst₂ : NonUnitalSemiring R) :
+    inst₁ = inst₂ ↔
+      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+      ∧ (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
+  ⟨fun h ↦ by constructor <;> (intros; congr), And.elim (ext · ·)⟩
+
+end NonUnitalSemiring
+
+/-! ### NonAssocSemiring -/
+/- TODO consider relocating these lemmas. -/
+@[ext] theorem AddMonoidWithOne.ext ⦃inst₁ inst₂ : AddMonoidWithOne R⦄
```

Comment:
Would it work to move these results to `Mathlib/Algebra/Group/WithOne/Basic.lean`? The same for `AddGroupWithOne`.

Reply:
`Mathlib/Algebra/Group/WithOne` appears to be about an operator `WithOne` which adjoins a new zero/one to a type/algebraic structure, rather than about algebraic structures which already have one.

The `AddMonoidWithOne` etc. about which we have extra ext lemmas here are all defined in `Mathlib/Data/Nat/Cast/Defs.lean` or `Mathlib/Data/Int/Cast/Defs.lean`. As to where that means their ext lemmas should be placed, :shrug:. There is a resemblance in name to the classes of `Mathlib/Algebra/GroupWithZero`...